### PR TITLE
Add support for `wasi:config/runtime` virtualization

### PR DESCRIPTION
**Pull request by scothis**
The [`wasi:config/runtime`](https://github.com/webassembly/wasi-runtime-config) interface is a way to pass configuration to other modules with a standard interface. Similar to `wasi:cli/environment` the core data model is a set of key-value pairs.

The proposed API and CLI support in WASI-Virt mirrors the support for environment variables that currently exists.

The config runtime proposal is currently stage 2 and does not a have a release. Because of that, I understand it may not be desirable to land support inside wasi-virt at this time. However, if there is a willingness to move forward, I can refine this PR to DRY-up the similarities between env and config, and flesh out docs, tests, etc.

### CODE REVIEW

**Comment by guybedford** [comment]
Thanks so much for looking into this. Nice to see it implemented as a separate adapter component, to verify - if it's not used, are we able to DCE the adapter similarly to the environment subsystem?

I will aim to review in the next few days.

**Comment by scothis** [comment]
> to verify - if it's not used, are we able to DCE the adapter similarly to the environment subsystem?

Currently following the same pattern used for `wasi:cli/environment`.

The config adapter is only injected into the module when the config option is enabled.

https://github.com/bytecodealliance/WASI-Virt/pull/82/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R156-R158

...and stripped out when not active

https://github.com/bytecodealliance/WASI-Virt/pull/82/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R229-R233

Something that needs to be handled better is `--allow-all`, which adds the `wasi:config/runtime` import to the resulting module even if the wrapped module doesn't import that interface. Hosts that don't provide the interface (like `wasi:http/proxy` and `wasi:cli/command`) will no longer be able to satisfy the module imports.

Have you thought about limiting the imported interfaces on the module to not exceed the imports on the wrapped module? I can't think of a reason to expose an import that will never be consumed.

**Comment by guybedford** [comment]
> Have you thought about limiting the imported interfaces on the module to not exceed the imports on the wrapped module? I can't think of a reason to expose an import that will never be consumed.

This makes a lot of sense - a DCE of the virtual module to the known module it is adapting, when the virtual module is not being generated on its own but in this composition.

Would you be interested in doing that here or are you thinking of a follow-up perhaps?

**Comment by scothis** [comment]
A separate PR that lands first makes sense. I'll be offline for a few days, but can take a look next week.

**Review by guybedford — COMMENTED** [review]
> Something that needs to be handled better is --allow-all, which adds the wasi:config/runtime import to the resulting module even if the wrapped module doesn't import that interface. Hosts that don't provide the interface (like wasi:http/proxy and wasi:cli/command) will no longer be able to satisfy the module imports.

I wonder if we should change this to support "WASI sets", something like `--allow-all` `--allow-core`? Where this would be in the `all` set but not the `core` set?

This PR looks great to me though to land, we just need to ensure we have a test case that covers at least standard configuration use cases. Let me know if I can assist further at all.

File: tests/virt.rs [hunk 1]
```diff
@@ -396,3 +438,65 @@ fn collect_component_imports(component_bytes: Vec<u8>) -> Result<Vec<String>> {
...
+        >,
+    >
+    where
+        'life0: 'async_trait,
+        Self: 'async_trait,
+    {
+        let result = Ok(self.props.clone());
+        let result = std::future::ready(result);
+        let result = Box::new(result);
+        let result = Pin::new(result);
```
Comment (scothis) [inline tests/virt.rs:498]: there's probably a more idiomatic way to construct the result. Async rust is a bit beyond me at the moment.

**Comment by scothis** [comment]
Rebased on main and added tests/doc.

File: src/bin/wasi-virt.rs [hunk 2]
```diff
@@ -58,6 +58,15 @@ struct Args {
     #[arg(short, long, use_value_delimiter(true), value_name("ENV=VAR"), value_parser = parse_key_val::<String, String>, help_heading = "Env")]
     env: Option<Vec<(String, String)>>,
 
+    // CONFIG
+    /// Allow unrestricted access to host configuration properties, or to a comma-separated list of property names.
```
Comment (guybedford) [inline src/bin/wasi-virt.rs:62]: Perhaps let's be explicit that this is the `runtime-config` system by including this term in the help?

File: README.md [hunk 3]
```diff
@@ -82,6 +83,19 @@ wasi-virt component.wasm -e CUSTOM=VAR --allow-env -o virt.wasm
 wasi-virt component.wasm -e CUSTOM=VAR --allow-env=SOME,ENV_VARS -o virt.wasm
 ```
 
+### Config
```
Comment (guybedford) [inline README.md:86]: Should this be `Runtime Config`?

**Review by guybedford — APPROVED** [review]
This seems to be the right approach to me. If you can confirm it's definitely not being pulled in for other subsystems I think we're good to land.

**Comment by scothis** [comment]
> confirm it's definitely not being pulled in for other subsystems

I'm exactly sure what you're looking for, so here's a few examples:

### no host access

```sh
wasi-virt -o virt.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  export wasi:cli/environment@0.2.0;
  export wasi:config/runtime@0.2.0-draft;
  export wasi:cli/exit@0.2.0;
  export wasi:random/insecure-seed@0.2.0;
  export wasi:random/insecure@0.2.0;
  export wasi:random/random@0.2.0;
  export wasi:io/error@0.2.0;
  export wasi:io/poll@0.2.0;
  export wasi:io/streams@0.2.0;
  export wasi:clocks/monotonic-clock@0.2.0;
  export wasi:clocks/wall-clock@0.2.0;
  export wasi:sockets/network@0.2.0;
  export wasi:sockets/ip-name-lookup@0.2.0;
  export wasi:sockets/tcp@0.2.0;
  export wasi:sockets/udp@0.2.0;
  export wasi:sockets/instance-network@0.2.0;
  export wasi:sockets/tcp-create-socket@0.2.0;
  export wasi:sockets/udp-create-socket@0.2.0;
  export wasi:http/types@0.2.0;
  export wasi:http/incoming-handler@0.2.0;
  export wasi:http/outgoing-handler@0.2.0;
  export wasi:cli/stdin@0.2.0;
  export wasi:cli/stdout@0.2.0;
  export wasi:cli/stderr@0.2.0;
  export wasi:cli/terminal-input@0.2.0;
  export wasi:cli/terminal-output@0.2.0;
  export wasi:cli/terminal-stdin@0.2.0;
  export wasi:cli/terminal-stdout@0.2.0;
  export wasi:cli/terminal-stderr@0.2.0;
  export wasi:filesystem/types@0.2.0;
  export wasi:filesystem/preopens@0.2.0;
}
```

### stdio only

```sh
wasi-virt --allow-stdio -o virt.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:io/error@0.2.0;
  import wasi:io/poll@0.2.0;
  import wasi:io/streams@0.2.0;
  import wasi:cli/stdin@0.2.0;
  import wasi:cli/stdout@0.2.0;
  import wasi:cli/stderr@0.2.0;
  import wasi:cli/terminal-input@0.2.0;
  import wasi:cli/terminal-output@0.2.0;
  import wasi:cli/terminal-stdin@0.2.0;
  import wasi:cli/terminal-stdout@0.2.0;
  import wasi:cli/terminal-stderr@0.2.0;

  export wasi:cli/environment@0.2.0;
  export wasi:config/runtime@0.2.0-draft;
  export wasi:cli/exit@0.2.0;
  export wasi:random/insecure-seed@0.2.0;
  export wasi:random/insecure@0.2.0;
  export wasi:random/random@0.2.0;
  export wasi:io/error@0.2.0;
  export wasi:io/poll@0.2.0;
  export wasi:io/streams@0.2.0;
  export wasi:clocks/monotonic-clock@0.2.0;
  export wasi:clocks/wall-clock@0.2.0;
  export wasi:sockets/network@0.2.0;
  export wasi:sockets/ip-name-lookup@0.2.0;
  export wasi:sockets/tcp@0.2.0;
  export wasi:sockets/udp@0.2.0;
  export wasi:sockets/instance-network@0.2.0;
  export wasi:sockets/tcp-create-socket@0.2.0;
  export wasi:sockets/udp-create-socket@0.2.0;
  export wasi:http/types@0.2.0;
  export wasi:http/incoming-handler@0.2.0;
  export wasi:http/outgoing-handler@0.2.0;
  export wasi:cli/stdin@0.2.0;
  export wasi:cli/stdout@0.2.0;
  export wasi:cli/stderr@0.2.0;
  export wasi:cli/terminal-input@0.2.0;
  export wasi:cli/terminal-output@0.2.0;
  export wasi:cli/terminal-stdin@0.2.0;
  export wasi:cli/terminal-stdout@0.2.0;
  export wasi:cli/terminal-stderr@0.2.0;
  export wasi:filesystem/types@0.2.0;
  export wasi:filesystem/preopens@0.2.0;
}
```

### runtime config only

```sh
wasi-virt --allow-config -o virt.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:config/runtime@0.2.0-draft;

  export wasi:cli/environment@0.2.0;
  export wasi:config/runtime@0.2.0-draft;
  export wasi:cli/exit@0.2.0;
  export wasi:random/insecure-seed@0.2.0;
  export wasi:random/insecure@0.2.0;
  export wasi:random/random@0.2.0;
  export wasi:io/error@0.2.0;
  export wasi:io/poll@0.2.0;
  export wasi:io/streams@0.2.0;
  export wasi:clocks/monotonic-clock@0.2.0;
  export wasi:clocks/wall-clock@0.2.0;
  export wasi:sockets/network@0.2.0;
  export wasi:sockets/ip-name-lookup@0.2.0;
  export wasi:sockets/tcp@0.2.0;
  export wasi:sockets/udp@0.2.0;
  export wasi:sockets/instance-network@0.2.0;
  export wasi:sockets/tcp-create-socket@0.2.0;
  export wasi:sockets/udp-create-socket@0.2.0;
  export wasi:http/types@0.2.0;
  export wasi:http/incoming-handler@0.2.0;
  export wasi:http/outgoing-handler@0.2.0;
  export wasi:cli/stdin@0.2.0;
  export wasi:cli/stdout@0.2.0;
  export wasi:cli/stderr@0.2.0;
  export wasi:cli/terminal-input@0.2.0;
  export wasi:cli/terminal-output@0.2.0;
  export wasi:cli/terminal-stdin@0.2.0;
  export wasi:cli/terminal-stdout@0.2.0;
  export wasi:cli/terminal-stderr@0.2.0;
  export wasi:filesystem/types@0.2.0;
  export wasi:filesystem/preopens@0.2.0;
}
```

### all

```sh
wasi-virt --allow-all -o virt.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:cli/environment@0.2.0;
  import wasi:config/runtime@0.2.0-draft;
  import wasi:io/error@0.2.0;
  import wasi:io/poll@0.2.0;
  import wasi:io/streams@0.2.0;
  import wasi:clocks/monotonic-clock@0.2.0;
  import wasi:sockets/network@0.2.0;
  import wasi:sockets/ip-name-lookup@0.2.0;
  import wasi:sockets/tcp@0.2.0;
  import wasi:sockets/udp@0.2.0;
  import wasi:http/types@0.2.0;
  import wasi:http/outgoing-handler@0.2.0;
  import wasi:cli/stdin@0.2.0;
  import wasi:cli/stdout@0.2.0;
  import wasi:cli/stderr@0.2.0;
  import wasi:cli/terminal-input@0.2.0;
  import wasi:cli/terminal-output@0.2.0;
  import wasi:cli/terminal-stdin@0.2.0;
  import wasi:cli/terminal-stdout@0.2.0;
  import wasi:cli/terminal-stderr@0.2.0;
  import wasi:clocks/wall-clock@0.2.0;
  import wasi:filesystem/types@0.2.0;
  import wasi:filesystem/preopens@0.2.0;

  export wasi:cli/environment@0.2.0;
  export wasi:config/runtime@0.2.0-draft;
  export wasi:io/error@0.2.0;
  export wasi:io/poll@0.2.0;
  export wasi:io/streams@0.2.0;
  export wasi:clocks/monotonic-clock@0.2.0;
  export wasi:sockets/ip-name-lookup@0.2.0;
  export wasi:sockets/tcp@0.2.0;
  export wasi:sockets/udp@0.2.0;
  export wasi:http/types@0.2.0;
  export wasi:http/outgoing-handler@0.2.0;
  export wasi:cli/stdin@0.2.0;
  export wasi:cli/stdout@0.2.0;
  export wasi:cli/stderr@0.2.0;
  export wasi:cli/terminal-input@0.2.0;
  export wasi:cli/terminal-output@0.2.0;
  export wasi:cli/terminal-stdin@0.2.0;
  export wasi:cli/terminal-stdout@0.2.0;
  export wasi:cli/terminal-stderr@0.2.0;
  export wasi:filesystem/types@0.2.0;
  export wasi:filesystem/preopens@0.2.0;
}
```

### composed

The same component is wrapped for each of the examples below.

`component.wasm`:
```wit
package root:component;

world root {
  import wasi:logging/logging;
  import wasi:config/runtime@0.2.0-draft;

  export wasi:cli/run@0.2.0;
}
```

#### no host access

```sh
wasi-virt -o virt.wasm component.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:logging/logging;

  export wasi:cli/run@0.2.0;
}
```

#### runtime config only

```sh
wasi-virt --allow-config -o virt.wasm component.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:logging/logging;
  import wasi:config/runtime@0.2.0-draft;

  export wasi:cli/run@0.2.0;
}
```

#### all

```sh
wasi-virt --allow-all -o virt.wasm component.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:logging/logging;
  import wasi:config/runtime@0.2.0-draft;

  export wasi:cli/run@0.2.0;
}
```

#### all (without filtering added in #83)

```sh
wasi-virt --allow-all -o virt.wasm component.wasm
```
`virt.wasm`:
```wit
package root:component;

world root {
  import wasi:logging/logging;
  import wasi:cli/environment@0.2.0;
  import wasi:config/runtime@0.2.0-draft;
  import wasi:io/error@0.2.0;
  import wasi:io/poll@0.2.0;
  import wasi:io/streams@0.2.0;
  import wasi:clocks/monotonic-clock@0.2.0;
  import wasi:sockets/network@0.2.0;
  import wasi:sockets/ip-name-lookup@0.2.0;
  import wasi:sockets/tcp@0.2.0;
  import wasi:sockets/udp@0.2.0;
  import wasi:http/types@0.2.0;
  import wasi:http/outgoing-handler@0.2.0;
  import wasi:cli/stdin@0.2.0;
  import wasi:cli/stdout@0.2.0;
  import wasi:cli/stderr@0.2.0;
  import wasi:cli/terminal-input@0.2.0;
  import wasi:cli/terminal-output@0.2.0;
  import wasi:cli/terminal-stdin@0.2.0;
  import wasi:cli/terminal-stdout@0.2.0;
  import wasi:cli/terminal-stderr@0.2.0;
  import wasi:clocks/wall-clock@0.2.0;
  import wasi:filesystem/types@0.2.0;
  import wasi:filesystem/preopens@0.2.0;

  export wasi:cli/run@0.2.0;
}
```

**Review by guybedford — CHANGES_REQUESTED** [review]
Ahh, thanks for the summary, it seems this comment got overlooked still:

> I wonder if we should change this to support "WASI sets", something like --allow-all --allow-core? Where this would be in the all set but not the core set?

I think we should separate the concept of "wasi core" here somehow, via something like `--draft-proposals` or otherwise through the options to act as another layer to avoid this being output for standard workflows that don't specifically have access to this subsystem.

The risk specifically being that a virt component wouldn't run on Wasmtime since it would expect these imports under the flags in use today.

**Comment by scothis** [comment]
What if we drop it from `--allow-all` so that explicit opt-in is required. More options can be introduced in the future.

The interface would still show up on the exports of the virt component, but that should be safe.

**Comment by guybedford** [comment]
> The interface would still show up on the exports of the virt component, but that should be safe.

I believe we don't ever include subsystems that aren't explicitly included, are you sure this would still be the case? Cutting the bytes is important I think when features aren't needed.

**Comment by scothis** [comment]
> I believe we don't ever include subsystems that aren't explicitly included, are you sure this would still be the case? 

I followed the pattern from environment. When the subsystem isn't enabled the interface still exists but the functions are replaced.

https://github.com/bytecodealliance/WASI-Virt/blob/c4512614d0bbe50044b93577099232c1701b06e2/src/lib.rs#L273-L278
https://github.com/bytecodealliance/WASI-Virt/blob/c4512614d0bbe50044b93577099232c1701b06e2/src/virt_env.rs#L233-L250

> Cutting the bytes is important I think when features aren't needed.

I generated a set of virt components with this branch and from [main@c451261](https://github.com/bytecodealliance/WASI-Virt/commit/c4512614d0bbe50044b93577099232c1701b06e2). The latent capability adds 3K over the equivlent., actually using `--allow-config` adds 1K. The delta between `--allow-config` and `--allow-env` is -1K. The delta between `--allow-config` and `--allow-all` is 88K.

```
219K branch-allow-all.wasm
131K branch-allow-config.wasm
132K branch-allow-env.wasm
130K branch-none.wasm
216K main-allow-all.wasm
130K main-allow-env.wasm
127K main-none.wasm
```

**Comment by guybedford** [comment]
Thanks for confirming - so the exports are still written even when the subsystem is unused, but that isn't an issue for compatibility because exports are always optional for importers. And from a size perspective, we get exactly the same DCE situation as for env. The case where an unused subsystem generates exports are always stub exports anyway and this is the standard approach already, and this PR fully implements strip and stub virt operations to align with that.

Appreciated for sharing the numbers as well to verify the DCE!

**Pull request merged**